### PR TITLE
Poison import if `qiskit-terra` detected

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This uses `importlib.metadata` to search for installations of `qiskit-terra` at runtime, and poison the import if one is detected. The two packages are incompatible in a way that we cannot express to `pip` without having poisoned the `qiskit-terra` package as a whole, which is something we're less willing to do in order to keep it easier to install old versions of Qiskit-dependent software, and to keep the Qiskit 1.0 packaging story clean for the coming years.

`qiskit>=1.0` and `qiskit-terra<1.0` both attempt to install files in the same directories, so we can't do the detection by looking at the Qiskit version; we'll likely get whichever was installed second if they've both been installed to `site-packages`, or otherwise, whichever comes first on the search path.  This uses the distribution component of `importlib.metadata` to detect installed _distributions_ rather than imported files, allowing us to distinguish `qiskit` from `qiskit-terra`.

### Details and comments

See #11617 for a similar PR made to the 0.45 series (which I propose we release as 0.45.3) and to be ported to the 0.46 branch.  The reason I think we also need the 0.45.x and 0.46.x series to have this error is in case they get accidentally installed _after_ Qiskit 1.0 (such as during a subsequent `pip install` command), and the import system resolves `qiskit.__file__` to the `qiskit-terra` `__init__.py` rather than the Qiskit 1.0 init.

A similar false-positive to the one I mentioned on #11617 applies here too, but is probably a fair amount more likely if a developer still has a `qiskit_terra.egg-info` directory in their path, such as in the repo root of Qiskit.  The remedy is to delete that directory; it's not part of the actual installation (and tbh, I'm not sure why pip/setuptools drops it there given that `pip uninstall` doesn't clear it up).

### CODE REVIEW

**Comment by jakelishman** [comment]
Given that this 1.0-branch poisoning is a bit more likely to trip up developers (the conditions needed for a false positive are _much_ simpler), perhaps it might be worth putting in an existence check for a file that's known to only exist in `qiskit-terra` as well.  For example, if `qiskit/namespace.py` or `qiskit/tools/events/pubsub.py` fail to import, then we're _almost_ certainly in false-positive land, and that's a fairly direct test of "is this environment in a conflicted state?".

**Comment by jakelishman** [comment]
The docs failure here is actually the poisoning working correctly; the problem is that `tox` builds the environment by installing the environment dependencies _first_, and only then building and installing the package.  Because of the bi-directional dependency from Aer, that's causing the docs build to be in an invalid environment, which this PR just found for us.

**Comment by jakelishman** [comment]
I've made #11621 that is one way of fixing the tox environments, and for now I've cherry-picked it into this branch to see if there's any further problems lurking.

*edit*: this PR now also contains #11625, which is fixes for more bugs in Qiskit 1.0 found by the fix of #11621.  I'm just cherry-picking in those commits to check this PR ahead of time; it should merge last in the chain.

**Comment by jakelishman** [comment]
Now rebased over #11621 and #11625, but I'll just mark on hold for the same reason as #11617: I need to write a migration guide and tips for setting up environments during the transition, then short-link to it in the error message.

**Comment by jakelishman** [comment]
Qiskit/documentation#710 should be merged and live now before any release would actually land.

File: qiskit/__init__.py
Comment (mtreinish) [inline qiskit/__init__.py:33]: This is a clever trick relying on the 1.0 removals to determine if the terra import is breaking anything or not.

File: qiskit/__init__.py
Comment (mtreinish) [inline qiskit/__init__.py]: Can you apply the same suggestions I had for this in #11617 so the error message is consistent?

File: qiskit/__init__.py
Comment (jakelishman) [reply qiskit/__init__.py:33]: Thanks!

Will (@wshanks) alluded to a different case we might want to catch / improve the documentation to handle over in https://github.com/Qiskit/documentation/pull/710#discussion_r1472323844 in addition to this.  If he's got time to find the repro steps he mentions I can see if there's anything we might want to do in the import mechanism.

I don't want it to get too complex, though - I'd rather have a couple of (minor) false negatives for devs than slow imports or false positives for all users

File: qiskit/__init__.py
Comment (jakelishman) [reply qiskit/__init__.py]: Yep, done in e1da911268.

**Review by mtreinish — APPROVED** [review]
This LGTM, if there is another edge case we need to detect here we can always expand change the check condition down the road. The one I'm thinking about is with editable installs and qiskit<0.43 leaving bits of `qiskit.providers` in an odd state, but I haven't though throught how we would detect that short of `import qiskit` not even working.

This will catch the vast majority of the common invalid install cases so lets move forward with this for now.

File: qiskit/__init__.py
Comment (wshanks) [reply qiskit/__init__.py:33]: I wasn't able to reproduce it by installing 0.25->0.45->main, so I don't think we need to worry about it. Either it takes an intricate set of steps to reproduce or I was just confused about something when I saw it.

Comment (jakelishman) [reply qiskit/__init__.py:33]: Ok cool, thanks Will.  If you do come across another case, let us know - it's possible for us to update this check without much pain (though I am _really_ nervous about how I've been deliberately breaking `import`!)

**Pull request merged**
